### PR TITLE
Add attributes to TOMLDecodeError

### PR DESCRIPTION
**Pull request by hukkin**
Adds the same attributes to `TOMLDecodeError` that `JSONDecodeError` has.

Created a discussion about this https://discuss.python.org/t/69546

edit: and a cpython issue https://github.com/python/cpython/issues/126175

### CODE REVIEW

File: src/tomli/_parser.py
Comment (nineteendo) [inline src/tomli/_parser.py:68]: Shouldn't we only deprecate `TOMLDecodeError(msg, doc, pos)`? `TOMLDecodeError()` and `TOMLDecodeError(msg)` are fine.

File: src/tomli/_parser.py
Comment (nineteendo) [inline src/tomli/_parser.py:80]: I would set the new instance variables to None here.

Comment (hukkin) [reply src/tomli/_parser.py:80]: What problems could this cause? If we don't set the variables, we can promise non-None types for the attributes. And since the attributes never were there when using positional args in `__init__` this shouldn't break anything.

File: src/tomli/_parser.py
Comment (hukkin) [reply src/tomli/_parser.py:68]: If we want to make progress towards a state where we have `json.JSONDecodeError` signature and where the exception's attributes always have a meaningful non-None value, then what you suggest is not enough. If we want an end state where passing exactly three positional arguments results in a special object similar to passing (msg, doc, pos) keyword arguments, then you're right.

File: src/tomli/_parser.py
Comment (nineteendo) [reply src/tomli/_parser.py:80]: No, this is worse, because now you need to use `getattr()`, because the attributes might not even exist:

```python
msg = getattr(exc, "msg")
doc = getattr(exc, "doc")
pos = getattr(exc, "pos")
lineno = getattr(exc, "lineno")
colno = getattr(exc, "colno")
print(f"{msg}: line {lineno} column {colno} (char {pos})")
```

File: src/tomli/_parser.py
Comment (nineteendo) [reply src/tomli/_parser.py:68]: At the very least I would allow to pass the arguments positionally, this is too verbose:

```python
raise TOMLDecodeError(msg="Invalid statement", doc=src, pos=pos)
```

Comment (hukkin) [reply src/tomli/_parser.py:68]: Does verbosity matter if it stays within the library? Surely the "primary API" of this exception class is simply to get caught and have its attributes accessed, not being instantiated outside the library.

File: src/tomli/_parser.py
Comment (hukkin) [reply src/tomli/_parser.py:80]: The attributes will exist 100% of the time when raised by the library and caught by a consumer of it, though. You only need getattr in the obscure case someone created an instance themselves.

Comment (nineteendo) [reply src/tomli/_parser.py:80]: A linter will complain about this, and we also know that the attributes can't be None when raised by the library.

File: src/tomli/_parser.py
Comment (hukkin) [reply src/tomli/_parser.py:68]: Allowing arguments positionally results in a different `TOMLDecodeError("msg, "doc", 0).args` tuple so can theoretically break someone. Current state cannot.

Comment (nineteendo) [reply src/tomli/_parser.py:68]: The only motivation for not allowing to pass them positionally is allowing to pass a 3-tuple to args.
That's even more obscure, so I don't think there's even a benefit to doing this.

File: src/tomli/_parser.py
Comment (hukkin) [reply src/tomli/_parser.py:80]: Hmm, which linter will complain? Mypy and flake8 seem to accept this just fine, they run in CI.

I don't feel strongly about this, I'm happy to add the Nones if cpython codeowners prefer that. If the None case stays deprecated, perhaps I can even persuade typeshed to conveniently lie about the types never being None :)

File: src/tomli/_parser.py
Comment (hukkin) [reply src/tomli/_parser.py:68]: Sorry, could you clarify what you mean by "this" precisely? But yeah the motivation for many things in this PR is full backwards compatibility.

File: src/tomli/_parser.py
Comment (nineteendo) [reply src/tomli/_parser.py:80]: Hmm, they don't.

File: src/tomli/_parser.py
Comment (nineteendo) [reply src/tomli/_parser.py:68]: Something like this:

```python
class TOMLDecodeError(ValueError):
    def __init__(self, msg: str | None = None, doc: str | None = None, pos: int | None = None):
        if msg is None or doc is None or pos is None:
            warn("not passing msg, doc or pos is deprecated", DeprecationWarning, 2)
            if msg is None:
                super().__init__()
            else:
                super().__init__(msg)
        else:
            ...
            self.msg = msg
            self.doc = doc
            self.pos = pos
            self.lineno = lineno
            self.colno = colno
```

Comment (hukkin) [reply src/tomli/_parser.py:68]: But wouldn't that
- completely break passing 4+ arguments
- break `exc.args` when passing 2 or 3 arguments
- break `exc.args` when passing one arg that is None

while current state breaks none of these cases.

File: src/tomli/_parser.py
Comment (hukkin) [reply src/tomli/_parser.py:80]: I'm happy to know that, and sorry, my bad :smile_cat:

File: src/tomli/_parser.py
Comment (nineteendo) [reply src/tomli/_parser.py:68]: Yeah, in theory this not backwards compatible, but who would do this:

```python
>>> raise ValueError(1, 2, 3)
Traceback (most recent call last):
  File "<python-input-0>", line 1, in <module>
    raise ValueError(1, 2, 3)
ValueError: (1, 2, 3)
```

Unless we know someone is relying on this, we shouldn't make the API significantly worse. If it does break someone, let them raise an issue.

Also, I'm pretty sure that deprecating passing the arguments positionally isn't what encukou meant.

It might be better to deprecate it in a followup PR.

Comment (hukkin) [reply src/tomli/_parser.py:68]: > Yeah, in theory this not backwards compatible, but who would do this:

Haha, don't ask me. I question why anyone would make an instance of TOMLDecodeError no matter what the args are they're using. But I can accept that in the stdlib we should expect anything.

It'd feel weird to me to have a strict back compatibility policy and then subjectively assess case by case whether it can be dodged.

> we shouldn't make the API significantly worse

IMO the API that really matters to users is being able to catch the exception and access attributes, which we don't make any worse. Creating instances we don't make any worse either, we just add a deprecation warning (we can also decide not to), in case someone creates in a way that doesn't add meaningful attribute values. But if we eventually want that all created instances have those attribute values, then deprecation is the way.


> If it does break someone, let them raise an issue.

Isn't this what we want to avoid with the back compat policy?

> It might be better to deprecate it in a followup PR.

This, or not deprecating at all, are fine by me. I think cpython codeowners should decide.

Comment (nineteendo) [reply src/tomli/_parser.py:68]: > This, or not deprecating at all, are fine by me. I think cpython codeowners should decide.

Just create a PR, an then you'll know.

**Comment by encukou** [comment]
`JSONDecodeError` accepts positional arguments, maybe TOMLDecodeError should too to be more compatible?
Perhaps something like:
```py
class TOMLDecodeError(ValueError):
    """An error raised if a document is not valid TOML."""
    def __init__(self, msg: str = None, doc: str = None, pos: Pos = None, *args):
        if (
            args
            or not isinstance(msg, str)
            or isinstance(doc, str)
            or not isinstance(pos, Pos)
        ):
            warnings.warn(
                "Free-form arguments for TOMLDecodeError are deprecated. "
                "Please set 'msg', 'doc' and 'pos' arguments.",
                DeprecationWarning,
                stacklevel=2,
            )
            if pos is not None:
                args = pos, *args
            if doc is not None:
                args = doc, *args
            if msg is not None:
                args = msg, *args
            super().__init__(*args)
        else:
            lineno = doc.count("\n", 0, pos) + 1
            if lineno == 1:
                colno = pos + 1
            else:
                colno = pos - doc.rindex("\n", 0, pos)

            if pos >= len(doc):
                coord_repr = "end of document"
            else:
                coord_repr = f"line {lineno}, column {colno}"
            errmsg = f"{msg} (at {coord_repr})"
            super().__init__(self, errmsg)

            self.msg = msg
            self.doc = doc
            self.pos = pos
            self.lineno = lineno
            self.colno = colno
```
(Not sure if typechecker will allow `str` & `Pos` rather than `str|None` & `Pos|None`, but it would be nice to show the non-deprecated signature! For stdlib, *typeshed* can IMO list the new signature only.)

**Comment by hukkin** [comment]
I think that would break `exc.args` in cases where:

- 3 positional args of types (str, str, int) are given
- any of the first 3 positional args passed in is None

Notably, the changing of `exc.args` in these cases also changes the output of `str(exc)` and `repr(exc)`.

If this is acceptable, let's do it.

An alternative is what's currently in this PR, and allowing `JSONDecodeError`s signature (positional args included) after the deprecation period, which I think shouldn't ever break anyone, but has the downside that it doesn't immediately allow the final signature.

> Not sure if typechecker will allow str & Pos rather than str|None & Pos|None, but it would be nice to show the non-deprecated signature

I think this should be no issue at all. We can type annotate the non-deprecated signature, and `# type: ignore` the lines in `TOMLDecodeError.__init__` that violate this, and that should be it.

> typeshed can IMO list the new signature only

I agree.

**Comment by nineteendo** [comment]
> * any of the first 3 positional args passed in is None

You could use a private sentinel, but this is overkill, the backwards compatibility policy isn't that strict.

**Comment by hukkin** [comment]
I read [PEP 387](https://peps.python.org/pep-0387/) and my interpretation is it doesn't seem to be lenient in a way that's applicable here.

> Unless it is going through the deprecation process below, the behavior of an API must not change in an incompatible fashion between any two consecutive releases.

If a change is breaking, it seems very strict, no exceptions. Where it does seem lenient is interpreting whether or not a change is breaking:

> This PEP takes the perspective that “backwards incompatibility” means preexisting code ceases to comparatively function after a change. It is acknowledged that this is not a concrete definition, but the expectation is people in general understand what is meant by “backwards incompatibility”

If someone shows up and says we broke their `TOMLDecodeError("str", "str", 0)` instantiation, I don't think an argument can be made here that our change was back compatible.

I understand if in practice the PEP is applied pragmatically, and it's ok to break use cases that seem extremely rare or low importance, etc, but to me it seems like the PEP doesn't say it's allowed. Should it?

**Comment by encukou** [comment]
Yup, there's a judgment call.

The chances of someone using `("str", "str", 0)` or `("str", None)`, *and* using `args` to get the info out, seem *vanishingly* low. (And using a private sentinel rather than `None` would push that even lower.)

The original version of PR would break code that's not common, but it's *likely* to exist -- using one positional string (code copied from `tomllib` itself), or no arguments (lowest-effort way to mock the exception). Those should definitely keep working -- along with ones where there's no other reason to change them.
Here, we're trying to match `JSONDecodeError`, which is overwhelmingly called with positional arguments. Changing the `(msg, doc, pos)` case is, *IMO*, worth betting that no one uses this to fill in *args*.

**Comment by hukkin** [comment]
Thanks a lot for the reviews and comments!

Yeah, I agree this makes sense. I was mainly wondering whether making such judgment calls should be explicitly allowed by the PEP, as my interpretation was they may currently not be.

I've pushed another commit that includes a couple tests. As far as I'm aware, this revision doesn't break anything except the highly unlikely `TOMLDecodeError("str", "str", 0).args` attribute access (and consequently `str(e)` in the same case).

I'll make a sibling PR to cpython in the upcoming days.

File: src/tomli/_parser.py
Comment (nineteendo) [inline src/tomli/_parser.py:181]: Can't pass them positionally now?

Comment (hukkin) [reply src/tomli/_parser.py:181]: Oh forgot about that, yes we can now. I'll edit these, thanks!

File: src/tomli/_parser.py
Comment (AlexWaygood) [inline src/tomli/_parser.py:73]: If you wanted a type-safe sentinel that mypy is happy with, you could do something like this (the diff is relative to your PR branch):

```diff
diff --git a/src/tomli/_parser.py b/src/tomli/_parser.py
index e209f2e..cd35d9b 100644
--- a/src/tomli/_parser.py
+++ b/src/tomli/_parser.py
@@ -7,7 +7,7 @@ from __future__ import annotations
 from collections.abc import Iterable
 import string
 from types import MappingProxyType
-from typing import IO, Any, NamedTuple
+from typing import IO, Any, NamedTuple, final
 import warnings
 
 from ._re import (
@@ -52,7 +52,8 @@ BASIC_STR_ESCAPE_REPLACEMENTS = MappingProxyType(
 
 # Sentinel to be used as default arg during deprecation period of
 # TOMLDecodeError's free-form arguments
-_DEFAULT = object()
+@final
+class _DEFAULT: ...
 
 
 class TOMLDecodeError(ValueError):
@@ -68,9 +69,9 @@ class TOMLDecodeError(ValueError):
 
     def __init__(
         self,
-        msg: str = _DEFAULT,  # type: ignore[assignment]
-        doc: str = _DEFAULT,  # type: ignore[assignment]
-        pos: Pos = _DEFAULT,  # type: ignore[assignment]
+        msg: str | type[_DEFAULT] = _DEFAULT,
+        doc: str | type[_DEFAULT] = _DEFAULT,
+        pos: Pos | type[_DEFAULT] = _DEFAULT,
         *args: Any,
     ):
         if (
```

Comment (hukkin) [reply src/tomli/_parser.py:73]: Thanks, I actually did consider that, but since this is public API and passed on to users via `py.typed`, I think we may not want to communicate the sentinel type to users at all, as they should never use it.

Comment (AlexWaygood) [reply src/tomli/_parser.py:73]: Fair enough -- though it feels to me like it sort of creeps into the API anyway? It will still be visible as the default value if you call `help(TomlDecodeError)` or `inspect.signature(TomlDecodeError)`.

Using a class rather than an `object` instance also gives you a better repr for things like `help()` and `inspect.signature()`, so even if you don't want to adjust the type hint to make mypy happy, it might be worth considering. E.g. if you called the class `_DEPRECATED_DEFAULT`, you could get something like this:

```pycon
>>> import tomli
>>> help(tomli.TOMLDecodeError)
Help on class TOMLDecodeError in module tomli:

class TOMLDecodeError(builtins.ValueError)
 |  TOMLDecodeError(
 |      msg: 'str' = <class 'tomli._parser._DEPRECATED_DEFAULT'>,
 |      doc: 'str' = <class 'tomli._parser._DEPRECATED_DEFAULT'>,
 |      pos: 'Pos' = <class 'tomli._parser._DEPRECATED_DEFAULT'>,
 |      *args: 'Any'
 |  )
 |
```

Whereas currently on your branch you get this:

```pycon
>>> import tomli
>>> help(tomli.TOMLDecodeError)
Help on class TOMLDecodeError in module tomli:

class TOMLDecodeError(builtins.ValueError)
 |  TOMLDecodeError(
 |      msg: 'str' = <object object at 0x10075c880>,
 |      doc: 'str' = <object object at 0x10075c880>,
 |      pos: 'Pos' = <object object at 0x10075c880>,
 |      *args: 'Any'
 |  )
 |
```

Anyway -- just a thought! Feel free to toss the idea in the bin if you don't like it :-)

Comment (hukkin) [reply src/tomli/_parser.py:73]: Thanks for your insight! I didn't consider how this shows up on `help()`.

I do like how that reads, given that the word "deprecated" is in the sentinel's name.

I notice you used `typing.final`. Does that change anything here if the sentinel is private anyway?

Comment (AlexWaygood) [reply src/tomli/_parser.py:73]: > I notice you used `typing.final`. Does that change anything here if the sentinel is private anyway?

No, it doesn't really change anything here in practical terms.

_If_ you wanted to use the `str | type[_DEPRECATED_DEFAULT]` type hint, then it has some nice (but not really very important) properties:
- `str | type[_DEPRECATED_DEFAULT]` means "any string, or any subclass of the `_DEPRECATED_DEFAULT` class"
- But that's not really what we want here -- ideally we want a type hint that expresses "any string, or exactly the `_DEPRECATED_DEFAULT` class (excluding all proper subclasses of the class)"
- We can't get that exactly (it's inexpressable with type hints), but we can at least tell the type checker that the `_DEPRECATED_DEFAULT` class should never be subclassed by using the `@final` decorator. If we do that, then `type[_DEPRECATED_DEFAULT]` basically means the same thing as the type hint we'd _like_ but _can't have_, since for a class decorated with `@final` the set of possible subclasses of the class is exactly equal to a set that only contains the class itself.

So, it's a very minor thing I "like" to do in situations like this, but has very minimal practical benefits 😄

Comment (hukkin) [reply src/tomli/_parser.py]: Thanks Alex! I pushed the change.

**Review by encukou — APPROVED** [review]
> I was mainly wondering whether making such judgment calls should be explicitly allowed by the PEP, as my interpretation was they may currently not be.

Well, here I am essentially betting that no one, so far, wrote code that would be broken by this -- i.e. the “preexisting code” doesn't exist.
If I'm wrong, we'll need to revert.

**Pull request merged**
